eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs

```diff
@@ -0,0 +1,351 @@
+    "If there is no stake, accept iff threshold is zero"
+    ( \(re, rs, gas) ->
+        spoAccepted
+          @era
+          re {reStakePoolDistr = PoolDistr Map.empty (fromJust . toCompact $ Coin 100)}
+          rs
+          gas
+          `shouldBe` votingStakePoolThreshold @era rs (gasAction gas)
+          == SJust minBound
+    )
+
+allAbstainProp ::
+  forall era.
+  ( Arbitrary (PParamsHKD StrictMaybe era)
+  , Arbitrary (PParamsHKD Identity era)
+  , ConwayEraPParams era
+  ) =>
+  Spec
+allAbstainProp =
+  prop @((RatifyEnv era, RatifyState era, GovActionState era) -> Property)
+    "If all votes are abstain, accepted ratio is zero"
+    $ \(re, rs, gas) -> forAll
+      ( genTestData @era
+          (Ratios {yes = 0, no = 0, abstain = 50 % 100, alwaysAbstain = 50 % 100, noConfidence = 0})
+      )
+      $ \TestData {..} ->
+        spoAcceptedRatio
+          @era
+          re {reStakePoolDistr = distr, reDelegatees = delegatees, rePoolParams = poolParams}
+          gas {gasStakePoolVotes = votes}
+          (rs ^. rsEnactStateL . ensProtVerL)
+          `shouldBe` 0
+
+noVotesProp ::
+  forall era.
+  ( Arbitrary (PParamsHKD StrictMaybe era)
+  , Arbitrary (PParamsHKD Identity era)
+  , ConwayEraPParams era
+  ) =>
+  Spec
+noVotesProp =
+  prop @((RatifyEnv era, RatifyState era, GovActionState era) -> Property)
+    "If there are no votes, accepted ratio is zero"
+    $ \(re, rs, gas) -> forAll
+      ( genTestData @era
+          (Ratios {yes = 0, no = 0, abstain = 0, alwaysAbstain = 0, noConfidence = 0})
+      )
+      $ \TestData {..} ->
+        spoAcceptedRatio
+          @era
+          re {reStakePoolDistr = distr}
+          gas {gasStakePoolVotes = votes}
+          (rs ^. rsEnactStateL . ensProtVerL)
+          `shouldBe` 0
+
+allYesProp ::
+  forall era.
+  ( Arbitrary (PParamsHKD StrictMaybe era)
+  , Arbitrary (PParamsHKD Identity era)
+  , ConwayEraPParams era
+  ) =>
+  Spec
+allYesProp =
+  prop @((RatifyEnv era, RatifyState era, GovActionState era) -> Property)
+    "If all vote yes, accepted ratio is 1 (unless there is no stake) "
+    ( \(re, rs, gas) ->
+        forAll
+          ( genTestData @era
+              (Ratios {yes = 100 % 100, no = 0, abstain = 0, alwaysAbstain = 0, noConfidence = 0})
+          )
+          ( \TestData {..} ->
+              if fromCompact totalStake == Coin 0
+                then
+                  spoAcceptedRatio
+                    @era
+                    re {reStakePoolDistr = distr}
+                    gas {gasStakePoolVotes = votes}
+                    (rs ^. rsEnactStateL . ensProtVerL)
+                    `shouldBe` 0
+                else
+                  spoAcceptedRatio
+                    @era
+                    re {reStakePoolDistr = distr}
+                    gas {gasStakePoolVotes = votes}
+                    (rs ^. rsEnactStateL . ensProtVerL)
+                    `shouldBe` 1
```

Comment:
It is not easy to read that the two calls to `spoAcceptedRatio` are the same. Perhaps making a common reference will help in making it easier to understand.
```suggestion
              let acceptedRatio =
                    spoAcceptedRatio
                      @era 
                      re {reStakePoolDistr = distr}
                      gas {gasStakePoolVotes = votes}
                      (rs ^. rsEnactStateL . ensProtVerL)
               in if fromCompact totalStake == Coin 0
                     then acceptedRatio `shouldBe` 0
                     else acceptedRatio `shouldBe` 1
                
```